File: tree/BinaryTreeLevelOrderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <queue>

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (root == nullptr) return {};

        std::queue<TreeNode*> q;
        q.push(root);
        std::vector<std::vector<int>> order;
        while (!q.empty()) {
            std::vector<int> level;

            size_t levelSize = q.size();
            for (size_t i = 0; i < levelSize; i++) {
                auto node = q.front();
                q.pop();
                level.push_back(node->val);

                if (node->left != nullptr)  q.push(node->left);
                if (node->right != nullptr) q.push(node->right);
            }

            order.push_back(level);
        }

        return order;
    }
};

File: tree/binary_tree_level_order_traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        queue1 = []
        queue2 = []
        queue1.append(root)
        queue = queue1
        other_queue = queue2

        order = []
        while (queue):
            level = []
            while (queue):
                element = queue.pop(0)
                if not element: continue

                # adds level children to the other queue
                other_queue.append(element.left)
                other_queue.append(element.right)
                level.append(element.val)

            if (level):
                order.append(level.copy())
            level.clear()
            queue.clear()
            # swap queues
            queue, other_queue = other_queue, queue

        return order
